rust-runtime/aws-smithy-http/src/event_stream/input.rs

```diff
@@ -48,19 +48,20 @@ where
     }
 }
 
-/// Adapts a `Stream<SmithyMessageType>` to a signed `Stream<Bytes>` by using the provided
-/// message marshaller and signer implementations.
-///
-/// This will yield an `Err(SdkError::ConstructionFailure)` if a message can't be
-/// marshalled into an Event Stream frame, (e.g., if the message payload was too large).
-#[pin_project]
-pub struct MessageStreamAdapter<T, E> {
-    marshaller: Box<dyn MarshallMessage<Input = T> + Send + Sync>,
-    signer: Box<dyn SignMessage + Send + Sync>,
-    #[pin]
-    stream: Pin<Box<dyn Stream<Item = Result<T, BoxError>> + Send>>,
-    end_signal_sent: bool,
-    _phantom: PhantomData<E>,
+pin_project! {
+    /// Adapts a `Stream<SmithyMessageType>` to a signed `Stream<Bytes>` by using the provided
+    /// message marshaller and signer implementations.
+    ///
+    /// This will yield an `Err(SdkError::ConstructionFailure)` if a message can't be
+    /// marshalled into an Event Stream frame, (e.g., if the message payload was too large).
+    pub struct MessageStreamAdapter {
+        marshaller: Box<dyn MarshallMessage<Input = T> + Send + Sync>,
+        signer: Box<dyn SignMessage + Send + Sync>,
+        #[pin]
+        stream: Pin<Box<dyn Stream<Item = Result<T, BoxError>> + Send>>,
```

Comment:
Is pin projection actually needed here if the member is `Pin`? It seems like it shouldn't be, but I'm not 100% sure.

Reply:
I don't understand why it needs to be `Pin` but if I remove it then the compiler starts complaining about it not implementing some obtuse generic traits.

Reply:
@jdisanti is correct—the key here is using `.as_mut()` on the stream instead of using pin projection. 

Interestingly you actually need to keep `pin_project!` on the root (or you need to manually implement `Unpin` (which is safe (!!))